File: 907-KokoEatingBananas/907-KokoEatingBananas.cpp
// Last updated: 3/31/2026, 9:36:32 PM
#include <vector>
#include <algorithm>

class Solution {
public:
    bool kokoEats(const std::vector<int>& piles, int k, int h) {
        int koko = 0;
        for (auto p : piles) {
            koko += (p+k-1) / k; 
        }
        return koko <= h;
    }

    int minEatingSpeed(std::vector<int>& piles, int h) {
        int low = 1;
        int high = *std::max_element(piles.begin(), piles.end());

        while (low < high) {
            int mid = low + (high - low) / 2; 

            if (kokoEats(piles, mid, h)) {
                high = mid;
            } else {
                low = mid + 1; 
            }
        }

        return low; 
    }
};

File: 907-KokoEatingBananas/907-KokoEatingBananas.py
# Last updated: 3/31/2026, 9:36:28 PM
class Solution:
    def minEatingSpeed(self, piles: List[int], T: int) -> int:
        l, h = 1, max(piles)
        maxbanana = h
        while l<=h:
            mid = (h+l)//2
            tt = 0
            for i in range(len(piles)):
                tt += math.ceil(piles[i]/mid)
            if tt<=T:
                maxbanana = mid
                h = mid-1
            else:
                l = mid+1
        return maxbanana

